# Part 1. Equation of a Slime

How many late days are you using for this assignment?

In [82]:
# Imports section
import pandas as pd
import sklearn
import sklearn.model_selection
from sklearn.linear_model import LinearRegression, LogisticRegression
from sklearn.model_selection import cross_val_score, KFold
from sklearn.preprocessing import PolynomialFeatures
from sklearn.svm import SVC
from sklearn.neighbors import KNeighborsClassifier
from sklearn.neural_network import MLPClassifier
import numpy as np

## 1. Loading the dataset

In [42]:
# Using pandas load the dataset
data_set = pd.read_csv(r'C:\Users\deshn\Downloads\assignment-4-sklearn-for-machine-learning-<redacted>\assignment-4-sklearn-for-machine-learning-<redacted>\science_data_large.csv')
# Output the first 15 rows of the data
print(data_set.head(15))
# Display a summary of the table information (number of datapoints, etc.)
print(data_set.describe())

    Temperature °C  Mols KCL     Size nm^3
0              469       647  6.244743e+05
1              403       694  5.779610e+05
2              302       975  6.196847e+05
3              779       916  1.460449e+06
4              901        18  4.325726e+04
5              545       637  7.124634e+05
6              660       519  7.006960e+05
7              143       869  2.718260e+05
8               89       461  8.919803e+04
9              294       776  4.770210e+05
10             991       117  2.441771e+05
11             307       781  5.006455e+05
12             206        70  3.145200e+04
13             437       599  5.390215e+05
14             566        75  9.185271e+04
       Temperature °C     Mols KCL     Size nm^3
count     1000.000000  1000.000000  1.000000e+03
mean       500.500000   471.530000  5.086111e+05
std        288.819436   288.482872  4.474838e+05
min          1.000000     1.000000  1.611429e+01
25%        250.750000   226.750000  1.298267e+05
50%        500.500

## 2. Splitting the dataset

In [43]:
# Take the pandas dataset and split it into our features (X) and label (y)
X = data_set[['Temperature °C','Mols KCL']]
Y = data_set['Size nm^3']

# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
X_train,X_test,Y_train,Y_test = sklearn.model_selection.train_test_split(X,Y,test_size=0.1,random_state=42)
# For grading consistency use random_state=42 

## 3. Perform a Linear Regression

In [61]:
# Use sklearn to train a model on the training set
regression =LinearRegression().fit(X_train,Y_train)
# Create a sample datapoint and predict the output of that sample with the trained model
sample_point = [[405, 723]]
prediction=regression.predict(sample_point)
print('Predicted output for sample point: ', prediction)

# Report the score for that model using the default score function property of the SKLearn model, in your own words (markdown, not code) explain what the score means
score= regression.score(X_test,Y_test)
print('Score for the model: ', score)
# Extract the coefficients and intercept from the model and write an equation for your h(x) using LaTeX
coff= regression.coef_
inter= regression.intercept_
print('Coefficients: ', coff)
print('Intercept: ', inter)

Predicted output for sample point:  [688036.35090145]
Score for the model:  0.8552472077276095
Coefficients:  [ 866.14641337 1032.69506649]
Intercept:  -409391.47958340764


c:\Users\deshn\AppData\Local\Programs\Python\Python313\Lib\site-packages\sklearn\utils\validation.py:2739: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  warnings.warn(


# Explaination 
From the results above we can see that the score $(R^2)$ of the linear regression model is 0.8552472077276095 ~ 0.85.
The score $(R^2)$ ranges from 0 to 1. A value of 0 indicates that the model is not able to predict the data. A value of 1 indicates that the model is able to predict the data perfectly.

therefore the score $(R^2)$ is 0.8552472077276095 ~ 0.85 shows that the model is able to predict the data very well based on Temperature and Mols. The score indicates a high corelation between the inputs and target outputs, this says that the model will be effective in making predicting size in the given problem.

let the input variables be $x_1$ = temperature and $x_2$= Mols and the target variable be $h(x)$. then
$$h(x) = -409391 +866.15 x_1 +1032.7x_2$$

in the above equation -409390 is the intercept and 866.15 and 1032.7 are the coefficients of the input variables Temperature and Mols respectively. this equation gives the linear relationship between the input variables and the target variable.

Write the linear equation of a slime: (example equation: $E = mc^2$)

Report on score and explain meaning:

## 4. Use Cross Validation

In [84]:
# Use the cross_val_score function to repeat your experiment across many shuffles of the data
# For grading consistency use n_splits=5 and random_state=42

k_fold= KFold(n_splits=5, random_state=42, shuffle=True)
validation_score = cross_val_score(regression,X,Y,cv=k_fold)
print('Scores',validation_score)

# Report on their finding and their significance
mean_score = validation_score.mean()
std_score = validation_score.std()
print('Mean score:',mean_score)
print('Standard deviation:',std_score)

n_trials = 10  # Repeat the cross-validation 10 times
all_scores = []

for trial in range(n_trials):
    # Use cross_val_score to evaluate the model
    scores = cross_val_score(regression, X, Y, cv=k_fold)
    all_scores.append(scores)

# Convert scores to a NumPy array for easier analysis
all_scores = np.array(all_scores)

# Print scores for each trial
for i, scores in enumerate(all_scores):
    print(f'Trial {i + 1} Scores: {scores}')

# Calculate mean and standard deviation across trials
mean_scores = all_scores.mean(axis=0)
std_scores = all_scores.std(axis=0)

# Displaying results in a table format
results_df = pd.DataFrame({
    'Fold': np.arange(1, 5 + 1),
    'Mean Score': mean_scores,
    'Standard Deviation': std_scores
})

print("\nCross-validation Results:")
print(results_df)

Scores [0.86151889 0.82742341 0.87195173 0.88166206 0.85609101]
Mean score: 0.8597294202684646
Standard deviation: 0.01838773713930639
Trial 1 Scores: [0.86151889 0.82742341 0.87195173 0.88166206 0.85609101]
Trial 2 Scores: [0.86151889 0.82742341 0.87195173 0.88166206 0.85609101]
Trial 3 Scores: [0.86151889 0.82742341 0.87195173 0.88166206 0.85609101]
Trial 4 Scores: [0.86151889 0.82742341 0.87195173 0.88166206 0.85609101]
Trial 5 Scores: [0.86151889 0.82742341 0.87195173 0.88166206 0.85609101]
Trial 6 Scores: [0.86151889 0.82742341 0.87195173 0.88166206 0.85609101]
Trial 7 Scores: [0.86151889 0.82742341 0.87195173 0.88166206 0.85609101]
Trial 8 Scores: [0.86151889 0.82742341 0.87195173 0.88166206 0.85609101]
Trial 9 Scores: [0.86151889 0.82742341 0.87195173 0.88166206 0.85609101]
Trial 10 Scores: [0.86151889 0.82742341 0.87195173 0.88166206 0.85609101]

Cross-validation Results:
   Fold  Mean Score  Standard Deviation
0     1    0.861519        1.110223e-16
1     2    0.827423        

Result and findings conclusion about variability and mean and score
the mean score of 0.86 from cross validation means that acrosss all the different sets the model is able to explain 86% of the variance in the target variable. It is generally considered as good performance.

## 5. Using Polynomial Regression

In [64]:
# Using the PolynomialFeatures library perform another regression on an augmented dataset of degree 2
# Perform k-fold cross validation (as above)
ply_model = PolynomialFeatures(degree=2)
new_X=ply_model.fit_transform(X)

X_trainP, X_testP, Y_trainP, Y_testP = sklearn.model_selection.train_test_split(new_X, Y, test_size=0.1, random_state=42)
new_reg = LinearRegression()
new_reg.fit(X_trainP, Y_trainP)
sample= ply_model.transform([[22,1.5]])
predic= new_reg.predict(sample)
print('Predicted size from polynomial regression is: ',predic)

# cross Validation
cross_p= cross_val_score(new_reg, X_trainP, Y_trainP, cv=5)
print('cross validation score: ',cross_p)
print('Mean cross validation score: ',cross_p.mean())
# Report on the metrics and output the resultant equation as you did in Part 3.
score_p= new_reg.score(X_testP, Y_testP)
print('Score on test data: ',score_p)

coef_p= new_reg.coef_
inter_p= new_reg.intercept_
print('coefficient: ',coef_p)
print('intercept: ',inter_p)


Predicted size from polynomial regression is:  [330.06430552]
cross validation score:  [1. 1. 1. 1. 1.]
Mean cross validation score:  1.0
Score on test data:  1.0
coefficient:  [ 0.00000000e+00  1.20000000e+01 -1.27197088e-07  1.26458843e-11
  2.00000000e+00  2.85714287e-02]
intercept:  2.0479375962167978e-05


c:\Users\deshn\AppData\Local\Programs\Python\Python313\Lib\site-packages\sklearn\utils\validation.py:2739: UserWarning: X does not have valid feature names, but PolynomialFeatures was fitted with feature names
  warnings.warn(


Write the polynomial equation of a slime: (example equation: $E = mc^2$)

$$h(x)= 2.0479*10^{-5}+12x_1-1.271970*10^{-7}x_2+ 1.26458843*10^{-11}x_1x_2+2x_1^2+2.85714287*10^{-2}x_2^2$$

Report on the score and interpret:

We see that the score of the model is 1, which means that the model  captures all the variance in the dependant variable for the given data. This also means that the values the model predicted were exaclty the actual values of the dependant variable.
This can be a result of potential overfitting of the model.
Maybe the test set is very similar to the training set or is small. This could also mean that the model is capturing more than just the pattern, it is also capturing the noise in the data, making it complex.
As the cross validation scores are also 1 on training and validation sets it could mean a perfect fit (that model is working very well) or just overfitting.

# Part 2. Chronic Kidney Disease Prediction via Classification

Create code and markdown cells as needed to perform classification and report on your results

In [47]:
# Load the dataset. Then train and evaluate the classification models.

CKD_dataset = pd.read_csv(r'C:\Users\deshn\Downloads\assignment-4-sklearn-for-machine-learning-<redacted>\assignment-4-sklearn-for-machine-learning-<redacted>\ckd_feature_subset.csv')
print(CKD_dataset.head(15))
# Split the dataset into training and testing sets
XC= CKD_dataset.drop('Target_ckd', axis=1)
YC = CKD_dataset['Target_ckd']
X_trainC, X_testC, Y_trainC, Y_testC = sklearn.model_selection.train_test_split(XC,YC, test_size=0.1, random_state=42)

         age        bp      wbcc  appet_poor  appet_good      rbcc  Target_ckd
0   0.688312  0.333333  0.000000           1           0  0.000000           1
1   0.545455  0.333333  0.128319           1           0  0.305085           1
2   0.714286  0.500000  0.238938           1           0  0.186441           1
3   0.688312  0.333333  0.283186           0           1  0.338983           1
4   0.441558  0.333333  0.221239           1           0  0.220339           1
5   0.740260  0.833333  0.265487           0           1  0.355932           1
6   0.870130  0.333333  0.668142           0           1  0.237288           1
7   0.870130  0.500000  0.150442           0           1  0.372881           1
8   0.194805  0.666667  0.380531           0           1  0.305085           1
9   0.753247  0.833333  0.163717           0           1  0.220339           1
10  0.701299  0.166667  0.504425           1           0  0.152542           1
11  0.610390  0.666667  0.367257           1        

# Logistic Regression

In [48]:
log_reg= LogisticRegression(max_iter= 200)
log_reg.fit(X_trainC, Y_trainC)
sample_log= X_testC.iloc[0].values.reshape(1,-1)
sample_log_pred= log_reg.predict_proba(sample_log)
print('Probability of the sample is: ', sample_log_pred)
Score_log= log_reg.score(X_testC, Y_testC)
print('Score of the model is: ', Score_log)
coef_log= log_reg.coef_
print('Coefficients of the model are: ', coef_log)
intercept_log= log_reg.intercept_
print('Intercept of the model is: ', intercept_log)

Probability of the sample is:  [[0.91572244 0.08427756]]
Score of the model is:  0.6875
Coefficients of the model are:  [[ 1.30130388  1.36829176  1.48498229  1.53980848 -1.52457627 -2.50595136]]
Intercept of the model is:  [-0.44692539]


c:\Users\deshn\AppData\Local\Programs\Python\Python313\Lib\site-packages\sklearn\utils\validation.py:2739: UserWarning: X does not have valid feature names, but LogisticRegression was fitted with feature names
  warnings.warn(


# Support Vector Machines

In [49]:
mod_svc= SVC(probability=True)
mod_svc.fit(X_trainC, Y_trainC)
sample_svc= X_testC.iloc[0].values.reshape(1,-1)
sample_svc_pred= mod_svc.predict_proba(sample_svc)
print('Probability of sample is: ', sample_svc_pred)
score_svc= mod_svc.score(X_testC, Y_testC)
print('Score of SVC is: ', score_svc)


Probability of sample is:  [[0.99876568 0.00123432]]
Score of SVC is:  0.875


c:\Users\deshn\AppData\Local\Programs\Python\Python313\Lib\site-packages\sklearn\utils\validation.py:2739: UserWarning: X does not have valid feature names, but SVC was fitted with feature names
  warnings.warn(


# K-Nearest Neighbors

In [57]:
K_mod= KNeighborsClassifier(n_neighbors=5)
K_mod.fit(X_trainC, Y_trainC)
sample_k= X_testC.iloc[0].values.reshape(1,-1)
prediction_k = K_mod.predict_proba(sample_k)
print('Prediction for sample: ', prediction_k)
score_k= K_mod.score(X_testC, Y_testC)
print('Score for sample: ', score_k)


Prediction for sample:  [[1. 0.]]
Score for sample:  0.9375


c:\Users\deshn\AppData\Local\Programs\Python\Python313\Lib\site-packages\sklearn\utils\validation.py:2739: UserWarning: X does not have valid feature names, but KNeighborsClassifier was fitted with feature names
  warnings.warn(


In [ ]:
neural_mod= MLPClassifier(hidden_layer_sizes= (10,10), activation='relu', solver='adam', max_iter=1000, random_state=42)
neural_mod.fit(X_trainC, Y_trainC)
sample_neu= X_testC.iloc[0].values.reshape(1,-1)
pro =neural_mod.predict_proba(sample_neu)
score_neu= neural_mod.score(X_testC, Y_testC)
print('Score',score_neu)
print('Probability',pro)


Score 0.875
Probability [[0.9966103 0.0033897]]


c:\Users\deshn\AppData\Local\Programs\Python\Python313\Lib\site-packages\sklearn\utils\validation.py:2739: UserWarning: X does not have valid feature names, but MLPClassifier was fitted with feature names
  warnings.warn(


In [79]:
k_fold= KFold(n_splits=5, random_state=42, shuffle=True)
classifications = {'Logistic Regression':log_reg, 'Support Vector Machines':mod_svc,'K-Nearest Neighbors': K_mod, 'Neural Network':neural_mod}
table={}
for name, classification in classifications.items():
    validation_score = cross_val_score(classification,XC,YC,cv=k_fold)
    print('Scores',validation_score)
    mean_score = validation_score.mean()
    std_score = validation_score.std()
    table[name]={'Average':mean_score,'Standard Deviation':std_score}

the_table = pd.DataFrame(table).T
the_table.columns = ['Average','Standard Deviation']
print()
print(the_table)

Scores [0.80645161 0.93548387 0.77419355 0.83333333 0.93333333]
Scores [0.90322581 1.         0.87096774 0.9        0.96666667]
Scores [0.93548387 1.         0.87096774 0.86666667 0.96666667]
Scores [0.90322581 0.96774194 0.87096774 0.96666667 0.96666667]

                          Average  Standard Deviation
Logistic Regression      0.856559            0.066269
Support Vector Machines  0.928172            0.047601
K-Nearest Neighbors      0.927957            0.052440
Neural Network           0.935054            0.040466


## Results and Conclusion for Classification Experiments

From the table outputed above we can see that the neural network model seems to be the best performing model. the mean says that the model predicts the outcome with 93% accuracy and the standard deviation says that the model is consistant throughout the various splits of the data.a good model to use and is reliable.

# SVM
with the second highest mean score of 0.928 and second lowest standard deviation of 0.0476. The mean of the model indicates that it can accurately predict 92.82% of the outcomes. The standard deviation indicates that the model is consistant throughout the various splits of the data and is therefore a good model to use and is reliable.

# KNN
with the third highest mean score of 0.927 and third lowest standard deviation of 0.052. The mean of the model indicates that it can accurately predict 92.7% of the outcomes. The standard deviation indicates that the model is consistant throughout the various splits of the data but not as much as SVM and is therefore a less reliable.

# Logistic Regression
with the mean of 0.85 and the standard deviation of 0.066. The mean of the model indicates that it can accurately predict 85% of the outcomes. The standard deviation indicates that the model is not as consistant throughout the various splits of the data and is therefore a not agood model to use and is as reliable.

In [80]:
# Experiments with Neural Network.

neu_table= {}
config = [((50,),'relu','adam'),((20,10),'tanh','adam'),((5,5,5),'logistic','sgd')]
num= 1
for h, a,s in config:
    neural_mod= MLPClassifier(hidden_layer_sizes=h, activation=a, solver=s, max_iter=1000, random_state=42)
    neural_mod.fit(X_trainC, Y_trainC)
    sample_neu= X_testC.iloc[0].values.reshape(1,-1)
    pro =neural_mod.predict_proba(sample_neu)
    score_neu= neural_mod.score(X_testC, Y_testC)
    validation = cross_val_score(neural_mod, X_trainC, Y_trainC, cv=5)
    mean_n= validation.mean()
    std_n= validation.std()
    neu_table['configeration',num]= [mean_n,std_n, score_neu]
    num += 1
    
the_table = pd.DataFrame(neu_table).T
the_table.columns = ['Average','Standard Deviation','Score']
print()
print(the_table)


c:\Users\deshn\AppData\Local\Programs\Python\Python313\Lib\site-packages\sklearn\utils\validation.py:2739: UserWarning: X does not have valid feature names, but MLPClassifier was fitted with feature names
  warnings.warn(
c:\Users\deshn\AppData\Local\Programs\Python\Python313\Lib\site-packages\sklearn\utils\validation.py:2739: UserWarning: X does not have valid feature names, but MLPClassifier was fitted with feature names
  warnings.warn(



                  Average  Standard Deviation  Score
configeration 1  0.941270            0.030031  0.875
              2  0.955820            0.036347  0.875
              3  0.737302            0.012054  0.625


c:\Users\deshn\AppData\Local\Programs\Python\Python313\Lib\site-packages\sklearn\utils\validation.py:2739: UserWarning: X does not have valid feature names, but MLPClassifier was fitted with feature names
  warnings.warn(


## Results and Conclusion for Neural Network Experiments

From the table above we can see that the :
# configeration 1 :
has a mean of 0.94 which is very close to 1 meaning that the model is working well in finding out the underlying relationship in the data, the score of of 0.875 also suggests a good performance of the model across all the splits of data, with a standard deviation of 0.03 showing that the variability across the splits is also very low. So is consistant too.

# configeration 2 :
has a mean of 0.95 which is  better than configeration 1 and very close to 1 meaning that the model is working well in finding out the underlying relationship in the data, the score of of 0.875 is same as configeration 1 even though the mean was better suggests the performance of the model across all the splits of data is the same as configeration 1, with a standard deviation of 0.036 showing that the variability across the splits is also very low. So is consistant too but slightly less than configeration 1.

# configeration 3 :
has a mean of 0.73 which is the worst seen but still not really bad and close to 1 meaning that the model is working in finding out the underlying relationship in the data, the score of of 0.625 suggests the performance of the model across all the splits of data is not stable, with a standard deviation of 0.012 showing that the variability across the splits is very low, so it does not work that well for any of them. So is consistant too but slightly less than configeration 1.

# Conclusion :
So we can see that configeration 2 works better than the others because it has 20  and 10 neurons so is likely able to learn complex features better than configeration 1 which has only one layer with 50 neurons and configeration 3. 
Given the performace we see that tanh works really well for the data while the 3rd configerattion uses the logistic activation that leads to slow learning in deep networks.